src/components/calendar/bl-calendar.ts

```diff
@@ -0,0 +1,545 @@
+    });
+  }
+  get days() {
+    return [...Array(7).keys()].map(day => {
+      return {
+        name: new Date(0, 0, day).toLocaleString(this.locale, { weekday: "short" }),
+        value: day,
+      };
+    });
+  }
+  /**
+   * Fires when date selection changes
+   */
+  @event("bl-calendar") private _onBlCalendarChange: EventDispatcher<Date | Date[]>;
+  static get styles(): CSSResultGroup {
+    return [style];
+  }
+  getDayNumInAMonth(year: number, month: number) {
+    return new Date(year, month + 1, 0).getDate();
+  }
+  getWeekDayOfDate(year: number, month: number) {
+    return new Date(year, month, 1).getDay();
+  }
+  setPreviousCalendarView() {
+    this.clearRangePickerStyles();
+    if (this._calendarView === CALENDAR_VIEWS.days) {
+      if (this._calendarMonth === FIRST_MONTH_INDEX) {
+        this._calendarMonth = LAST_MONTH_INDEX;
+        this._calendarYear -= 1;
+      } else this._calendarMonth -= 1;
+    } else if (this._calendarView === CALENDAR_VIEWS.months) {
+      this._calendarYear -= 1;
+    } else if (this._calendarView === CALENDAR_VIEWS.years) {
+      const fromYear = this._calendarYears[0];
+
+      this._calendarYears = [];
+      for (let i = 12; i > 0; i--) {
+        this._calendarYears.push(fromYear - i);
+      }
+    }
+    if (this.type === CALENDAR_TYPES.range) {
+      this.setHoverClass();
+    }
+  }
+  setNextCalendarView() {
+    this.clearRangePickerStyles();
+    if (this._calendarView === CALENDAR_VIEWS.days) {
+      if (this._calendarMonth === LAST_MONTH_INDEX) {
+        this._calendarMonth = FIRST_MONTH_INDEX;
+        this._calendarYear += 1;
+      } else this._calendarMonth += 1;
+    } else if (this._calendarView === CALENDAR_VIEWS.months) {
+      this._calendarYear += 1;
+    } else if (this._calendarView === CALENDAR_VIEWS.years) {
+      const fromYear = this._calendarYears[11];
+
+      this._calendarYears = [];
+      for (let i = 1; i <= 12; i++) {
+        this._calendarYears.push(fromYear + i);
+      }
+    }
+    if (this.type === CALENDAR_TYPES.range) {
+      this.setHoverClass();
+    }
+  }
+
+  setCurrentCalendarView(view: CalendarView) {
+    if (this._calendarView !== view) {
+      this._calendarView = view;
+    } else this._calendarView = CALENDAR_VIEWS.days;
+    this.setHoverClass();
+  }
+
+  setMonthAndCalendarView(month: number) {
+    this._calendarMonth = month;
+    this._calendarView = CALENDAR_VIEWS.days;
+    if (this.type === CALENDAR_TYPES.range) {
+      this.setHoverClass();
+    }
+  }
+  setYearAndCalendarView(year: number) {
+    this._calendarYear = year;
+    this._calendarView = CALENDAR_VIEWS.days;
+    if (this.type === CALENDAR_TYPES.range) {
+      this.setHoverClass();
+    }
+  }
+
+  createCalendarYears() {
+    if (this._calendarYears.length === 0) {
+      for (let i = 4; i >= 0; i--) this._calendarYears.push(this._calendarYear - i);
+      for (let i = 1; i <= 7; i++) this._calendarYears.push(this._calendarYear + i);
+    }
+  }
```

Comment:
That's how I'd refactor it, I recommend finding a better name that explains the function something like `generateSurroundingYears`. 

```suggestion
  createCalendarYears() {
    if (this._calendarYears.length === 0) {
      this._calendarYears = Array.from({ length: 12 }, (_, index) => this._calendarYear - 4 + index);
    }
  }
```